### PR TITLE
feat(webapp): introduce openapi/swagger specification

### PR DESCRIPTION
**Pull request by bonomat**
It is a bit of manual work and boiler plates but once set up it's quite nice. 
I've enabled redoc and swagger UI for now but I guess one of them is enough

    
- OpenApi/Swagger can be found under: http://localhost:3001/swagger-ui
- Redoc can be found under: http://localhost:3001/redoc

Note: our authentication seems to be non-standard, and hence, it does not fit to the Swagger authentication. Imho we should fix this to have a fully working swagger client. Supported authentication schemes can be found [here](https://docs.rs/utoipa/latest/utoipa/openapi/security/enum.SecurityScheme.html) 

Helps with https://github.com/get10101/10101/issues/2284

### CODE REVIEW

File: webapp/src/api.rs [hunk 1]
```diff
@@ -93,18 +97,32 @@ pub async fn version() -> Json<Version> {
     })
 }
 
+#[utoipa::path(
+    get,
+    path = "/api/newaddress",
+    responses(
+       (status = 200, description = "Returns an unused on-chain address", body = String)
+)
```
Comment (luckysori) [inline webapp/src/api.rs:105]:
```suggestion
    )
```

File: webapp/src/api.rs [hunk 2]
```diff
@@ -93,18 +97,32 @@ pub async fn version() -> Json<Version> {
...
 pub struct Balance {
     on_chain: u64,
     off_chain: Option<u64>,
 }
 
+#[utoipa::path(
+get,
+path = "/api/balance",
+responses(
+(status = 200, description = "Returned on- and off-chain balance", body = Balance)
```
Comment (luckysori) [inline webapp/src/api.rs:123]:
```suggestion
(status = 200, description = "Returned on-chain and off-chain balance", body = Balance)
```

Comment (luckysori) [reply webapp/src/api.rs:123]: Formatting is a bit weird here and elsewhere.

Comment (luckysori) [reply webapp/src/api.rs:123]: Seems like it should read "Returns" not "Returned".

File: webapp/src/api.rs [hunk 3]
```diff
@@ -127,6 +145,13 @@ pub struct OnChainPayment {
     fee: Option<u64>,
 }
 
+#[utoipa::path(
+get,
+path = "/api/history",
+responses(
+(status = 200, description = "Retrieved on-chain payment history", body = [OnChainPayment])
```
Comment (luckysori) [inline webapp/src/api.rs:152]:
```suggestion
(status = 200, description = "Retrieves on-chain payment history", body = [OnChainPayment])
```

File: webapp/src/api.rs (hunk 2 above)
Comment (luckysori) [reply webapp/src/api.rs:123]: I guess we should be consistent. I'm not sure what the standard is for these `description`s.

File: webapp/src/api.rs [hunk 4]
```diff
@@ -188,37 +221,95 @@ pub async fn send_payment(
...
 pub struct NewOrderParams {
     #[serde(with = "rust_decimal::serde::float")]
     pub leverage: Decimal,
     #[serde(with = "rust_decimal::serde::float")]
     pub quantity: Decimal,
     pub direction: Direction,
-    #[serde(with = "bitcoin::amount::serde::as_sat::opt")]
-    pub coordinator_reserve: Option<Amount>,
-    #[serde(with = "bitcoin::amount::serde::as_sat::opt")]
-    pub trader_reserve: Option<Amount>,
```
Comment (luckysori) [inline webapp/src/api.rs:221]: 😞 That's a bit annoying.

File: webapp/src/auth.rs [hunk 5]
```diff
@@ -88,13 +89,21 @@ pub fn router() -> Router {
...
     use axum::response::IntoResponse;
     use axum::Json;
     use axum_login::AuthSession;
 
+    #[utoipa::path(
+        post,
+        path = "/api/login",
+        request_body = Credentials,
+        responses(
+        (status = 200, description = "If login was successfully", body = ())
```
Comment (luckysori) [inline webapp/src/auth.rs:104]:
```suggestion
        (status = 200, description = "If login was successful", body = ())
```

File: webapp/src/auth.rs [hunk 6]
```diff
@@ -115,11 +124,18 @@ mod post {
...
+pub mod get {
     use crate::api::AppError;
     use crate::auth::Backend;
     use axum_login::AuthSession;
 
+    #[utoipa::path(
+        get,
+        path = "/api/logout",
+        responses(
+        (status = 200, description = "If logout was successfully", body = ())
```
Comment (luckysori) [inline webapp/src/auth.rs:136]:
```suggestion
        (status = 200, description = "If logout was successful", body = ())
```

File: webapp/src/main.rs [hunk 7]
```diff
@@ -205,13 +214,81 @@ pub struct AppState {
...
+
     let router = Router::new()
         .route("/", get(index_handler))
         .route("/main.dart.js", get(main_dart_handler))
         .route("/flutter.js", get(flutter_js))
         .route("/index.html", get(index_handler))
         .route("/assets/*file", get(static_handler))
         .route("/api/version", get(version))
+        .merge(SwaggerUi::new("/swagger-ui").url("/api-docs/openapi.json", ApiDoc::openapi()))
+        .merge(Redoc::with_url("/redoc", ApiDoc::openapi()))
```
Comment (luckysori) [inline webapp/src/main.rs:290]: ❓Did you miss `/api-docs/` here?

**Review by luckysori — APPROVED** [review]
Nice! Just some style comments.

File: webapp/src/api.rs (hunk 2 above)
Comment (bonomat) [reply webapp/src/api.rs:123]: > Formatting is a bit weird here and elsewhere.

Rust formatter doesn't work here :/

> I guess we should be consistent. I'm not sure what the standard is for these descriptions.

Not sure neither 🙈

File: webapp/src/api.rs (hunk 4 above)
Comment (bonomat) [reply webapp/src/api.rs:221]: I know :( We could potentially wrap it into another type and then imple `ToSchema` but I found this was the easiest way.

File: webapp/src/main.rs (hunk 7 above)
Comment (bonomat) [reply webapp/src/main.rs:290]: No, Afaik `/swagger-ui` needs to find the specification under `/api-docs/openapi.json`. `redocs` has its own. We probably only need one of them.

**Pull request merged**